### PR TITLE
refactor and test: addressing PR 104 comments

### PR DESCRIPTION
**Pull request by kupermind**
- Optimizing a bit closed product Id calculation;
- Changing revert logic and names.

### CODE REVIEW

File: contracts/Depository.sol [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
-import "./interfaces/IErrorsTokenomics.sol";
-import "./interfaces/IGenericBondCalculator.sol";
-import "./interfaces/IToken.sol";
-import "./interfaces/ITokenomics.sol";
-import "./interfaces/ITreasury.sol";
+import {IErrorsTokenomics} from "./interfaces/IErrorsTokenomics.sol";
+import {IGenericBondCalculator} from "./interfaces/IGenericBondCalculator.sol";
+import {IToken} from "./interfaces/IToken.sol";
+import {ITokenomics} from "./interfaces/ITokenomics.sol";
+import {ITreasury} from "./interfaces/ITreasury.sol";
```
Comment (kupermind) [inline contracts/Depository.sol:8]: To get rid of solhint warnings

File: contracts/Depository.sol [hunk 2]
```diff
@@ -249,34 +249,30 @@ contract Depository is IErrorsTokenomics {
...
-            // The supply is greater than zero only if the product is active, otherwise it is instantly closed
+            // The supply is greater than zero only if the product is active, otherwise it is already closed
             if (supply > 0) {
                 // Refund unused OLAS supply from the product if it was not used by the product completely
                 ITokenomics(tokenomics).refundFromBondProgram(supply);
                 address token = mapBondProducts[productId].token;
                 delete mapBondProducts[productId];
 
-                positions[i] = true;
+                ids[numClosedProducts] = productIds[i];
```
Comment (kupermind) [inline contracts/Depository.sol:266]: closed Ids are recorded in the same loop

File: contracts/Depository.sol [hunk 3]
```diff
@@ -249,34 +249,30 @@ contract Depository is IErrorsTokenomics {
...
+        // Get the correct array size of closed product Ids
         closedProductIds = new uint256[](numClosedProducts);
-        uint256 numPos;
-        for (uint256 i = 0; i < numProducts; ++i) {
-            if (positions[i] ) {
-                closedProductIds[numPos] = productIds[i];
-                ++numPos;
-            }
+        for (uint256 i = 0; i < numClosedProducts; ++i) {
+            closedProductIds[i] = ids[i];
```
Comment (kupermind) [inline contracts/Depository.sol:275]: We need to copy to the correctly allocated array

File: contracts/Depository.sol [hunk 4]
```diff
@@ -303,28 +299,27 @@ contract Depository is IErrorsTokenomics {
...
-
-        // Check for the product vesting time
-        // Note that if the token or productId are invalid, the vesting will be zero by default and result in revert
-        uint256 vesting = product.vesting;
-        if (vesting == 0) {
-            revert ProductExpired(token, productId, vesting, block.timestamp);
+        // Check for the product supply, which is zero if the product was closed or never existed
+        uint256 supply = product.supply;
+        if (supply == 0) {
+            revert ProductClosed(productId);
```
Comment (kupermind) [inline contracts/Depository.sol:305]: When the product is closed or never existed, its supply is zero

**Review by DavidMinarsch — APPROVED** [review]

**Review by 77ph — APPROVED** [review]

**Pull request merged**